commit 6fd144e3a9664ea545d10008c48dceb6c4e9d38a
Author: Daniel Gustafsson <dgustafsson@postgresql.org>
Date:   Thu Feb 29 12:19:52 2024 +0100

    Fix integer underflow in shared memory debugging
    
    dsa_dump would print a large negative number instead of zero for
    segment bin 0.  Fix by explicitly checking for underflow and add
    special case for bin 0. Backpatch to all supported versions.
    
    Author: Ian Ilyasov <ianilyasov@outlook.com>
    Reviewed-by: Robert Haas <robertmhaas@gmail.com>
    Discussion: https://postgr.es/m/GV1P251MB1004E0D09D117D3CECF9256ECD502@GV1P251MB1004.EURP251.PROD.OUTLOOK.COM
    Backpatch-through: v12

diff --git a/src/backend/utils/mmgr/dsa.c b/src/backend/utils/mmgr/dsa.c
index a6b728ba9ff..b9e7f224d51 100644
--- a/src/backend/utils/mmgr/dsa.c
+++ b/src/backend/utils/mmgr/dsa.c
@@ -1111,13 +1111,17 @@ dsa_dump(dsa_area *area)
 	{
 		if (area->control->segment_bins[i] != DSA_SEGMENT_INDEX_NONE)
 		{
 			dsa_segment_index segment_index;
 
-			fprintf(stderr,
-					"    segment bin %zu (at least %d contiguous pages free):\n",
-					i, 1 << (i - 1));
+			if (i == 0)
+				fprintf(stderr,
+						"    segment bin %zu (no contiguous free pages):\n", i);
+			else
+				fprintf(stderr,
+						"    segment bin %zu (at least %d contiguous pages free):\n",
+						i, 1 << (i - 1));
 			segment_index = area->control->segment_bins[i];
 			while (segment_index != DSA_SEGMENT_INDEX_NONE)
 			{
 				dsa_segment_map *segment_map;
 
